// @mysten/sui/dist/cjs/jsonRpc/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var jsonRpc_exports = {};
__export(jsonRpc_exports, {
  JsonRpcError: () => import_errors.JsonRpcError,
  JsonRpcHTTPTransport: () => import_http_transport.JsonRpcHTTPTransport,
  SuiHTTPStatusError: () => import_errors.SuiHTTPStatusError,
  SuiHTTPTransportError: () => import_errors.SuiHTTPTransportError,
  SuiJsonRpcClient: () => import_client.SuiJsonRpcClient,
  isSuiJsonRpcClient: () => import_client.isSuiJsonRpcClient
});
module.exports = __toCommonJS(jsonRpc_exports);
var import_http_transport = require("./http-transport.js");
var import_client = require("./client.js");
var import_errors = require("./errors.js");
