// @shinami/clients/dist/cjs/aptos/aptos.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAptosClient = void 0;
/**
 * Copyright 2023-2024 Shinami Corp.
 * SPDX-License-Identifier: Apache-2.0
 */
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const endpoints_js_1 = require("./endpoints.js");
const region_js_1 = require("../region.js");
/**
 * A private function which infers the network enum from a Shinami access key
 * @param accessKey
 * @returns a Network enum type or undefined
 */
function inferNetworkFromKey(accessKey) {
    if (accessKey.includes("aptos_devnet")) {
        return ts_sdk_1.Network.DEVNET;
    }
    else if (accessKey.includes("aptos_testnet")) {
        return ts_sdk_1.Network.TESTNET;
    }
    else if (accessKey.includes("aptos_mainnet")) {
        return ts_sdk_1.Network.MAINNET;
    }
    else if (accessKey.includes("aptos_local")) {
        return ts_sdk_1.Network.LOCAL;
    }
    else {
        console.warn("Access key network qualifier not found, setting as undefined");
        return undefined;
    }
}
/**
 * Creates an Aptos client pointing to Shinami Node Service
 * @param accessKey Node access key, created in Shinami dashboard. Note that this determines what network you targeting
 * @param url Optional URL for fullnode. Defaults to Shinami's API
 * @param indexerUrl Optional URL for indexer. Defaults to Shinami's GraphQL server.
 * @returns Aptos client using Shinami Node service.
 */
function createAptosClient(accessKey, url = (0, region_js_1.inferRegionalValueFromAccessKey)(accessKey, endpoints_js_1.NodeRestUrls, (nodeRestUrls) => nodeRestUrls.us1), indexerUrl = (0, region_js_1.inferRegionalValueFromAccessKey)(accessKey, endpoints_js_1.NodeIndexerUrls, (nodeIndexerUrls) => nodeIndexerUrls.us1)) {
    const network = inferNetworkFromKey(accessKey);
    return new ts_sdk_1.Aptos(new ts_sdk_1.AptosConfig({
        fullnode: url,
        indexer: indexerUrl,
        clientConfig: {
            API_KEY: accessKey,
        },
        network: network,
    }));
}
exports.createAptosClient = createAptosClient;
