// @mysten/utils/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  DataLoader: () => import_dataloader.DataLoader,
  chunk: () => import_chunk.chunk,
  fromBase58: () => import_b58.fromBase58,
  fromBase64: () => import_b64.fromBase64,
  fromHex: () => import_hex.fromHex,
  promiseWithResolvers: () => import_with_resolver.promiseWithResolvers,
  toBase58: () => import_b58.toBase58,
  toBase64: () => import_b64.toBase64,
  toHex: () => import_hex.toHex
});
module.exports = __toCommonJS(index_exports);
var import_b58 = require("./b58.js");
var import_b64 = require("./b64.js");
var import_hex = require("./hex.js");
var import_chunk = require("./chunk.js");
var import_with_resolver = require("./with-resolver.js");
var import_dataloader = require("./dataloader.js");
