// @shinami/clients/dist/cjs/sui/endpoints.js
"use strict";
/**
 * Copyright 2023-2024 Shinami Corp.
 * SPDX-License-Identifier: Apache-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZkProverRpcUrls = exports.ZkWalletRpcUrls = exports.KeyRpcUrls = exports.WalletRpcUrls = exports.GasStationRpcUrls = exports.NodeWsUrls = exports.NodeRpcUrls = void 0;
const region_js_1 = require("../region.js");
exports.NodeRpcUrls = {
    us1: (0, region_js_1.createRegionalApiUrl)("us1", "sui", "node"),
    eu1: (0, region_js_1.createRegionalApiUrl)("eu1", "sui", "node"),
    apac1: (0, region_js_1.createRegionalApiUrl)("apac1", "sui", "node"),
};
exports.NodeWsUrls = {
    us1: (0, region_js_1.createRegionalApiUrl)("us1", "sui", "node", "wss"),
    eu1: (0, region_js_1.createRegionalApiUrl)("eu1", "sui", "node", "wss"),
    apac1: (0, region_js_1.createRegionalApiUrl)("apac1", "sui", "node", "wss"),
};
exports.GasStationRpcUrls = {
    us1: (0, region_js_1.createRegionalApiUrl)("us1", "sui", "gas"),
};
exports.WalletRpcUrls = {
    us1: (0, region_js_1.createRegionalApiUrl)("us1", "sui", "wallet"),
};
exports.KeyRpcUrls = {
    us1: (0, region_js_1.createRegionalApiUrl)("us1", "sui", "key"),
};
exports.ZkWalletRpcUrls = {
    us1: (0, region_js_1.createRegionalApiUrl)("us1", "sui", "zkwallet"),
};
exports.ZkProverRpcUrls = {
    us1: (0, region_js_1.createRegionalApiUrl)("us1", "sui", "zkprover"),
};
