// @shinami/clients/dist/cjs/sui/gas.js
"use strict";
/**
 * Copyright 2023-2024 Shinami Corp.
 * SPDX-License-Identifier: Apache-2.0
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildGaslessTransaction = exports.GasStationClient = exports.Fund = exports.SponsoredTransactionStatus = exports.SponsoredTransaction = void 0;
const transactions_1 = require("@mysten/sui/transactions");
const utils_1 = require("@mysten/sui/utils");
const superstruct_1 = require("superstruct");
const rpc_js_1 = require("../rpc.js");
const utils_js_1 = require("../utils.js");
const endpoints_js_1 = require("./endpoints.js");
const region_js_1 = require("../region.js");
/**
 * A fully sponsored transaction block.
 */
exports.SponsoredTransaction = (0, superstruct_1.object)({
    /**
     * Base64 encoded transaction bytes, including sponsor gas data.
     */
    txBytes: (0, superstruct_1.string)(),
    /**
     * Transaction digest. Can be used to identify this sponsored transaction.
     */
    txDigest: (0, superstruct_1.string)(),
    /**
     * Gas owner's signature.
     */
    signature: (0, superstruct_1.string)(),
    /**
     * Cost breakdown for the sponsor.
     */
    sponsorCost: (0, superstruct_1.optional)((0, superstruct_1.object)({
        computationCost: (0, superstruct_1.string)(),
        storageCost: (0, superstruct_1.string)(),
        storageRebate: (0, superstruct_1.string)(),
    })),
    /**
     * Expiration time of this sponsored transaction, in Unix epoch seconds.
     */
    expireAtTime: (0, superstruct_1.optional)((0, superstruct_1.number)()),
    /**
     * The last Sui epoch this sponsored transaction is valid for.
     */
    expireAfterEpoch: (0, superstruct_1.optional)((0, superstruct_1.string)()),
});
/**
 * The status of a sponsored transaction block.
 */
exports.SponsoredTransactionStatus = (0, superstruct_1.enums)([
    "IN_FLIGHT",
    "COMPLETE",
    "INVALID",
]);
/**
 * The fund information
 */
exports.Fund = (0, superstruct_1.object)({
    network: (0, superstruct_1.string)(),
    name: (0, superstruct_1.string)(),
    balance: (0, superstruct_1.number)(),
    inFlight: (0, superstruct_1.number)(),
    depositAddress: (0, superstruct_1.optional)((0, superstruct_1.string)()),
});
/**
 * Gas station RPC client.
 */
class GasStationClient extends rpc_js_1.ShinamiRpcClient {
    /**
     * @param accessKey Gas access key. Note that the access key also determines which network your
     *    transactions are targeting.
     * @param url Optional URL override.
     */
    constructor(accessKey, url = (0, region_js_1.inferRegionalValueFromAccessKey)(accessKey, endpoints_js_1.GasStationRpcUrls, (gasStationRpcUrls) => gasStationRpcUrls.us1)) {
        super(accessKey, url);
    }
    /**
     * Requests sponsorship for a transaction.
     * @param tx Gasless transaction.
     * @returns A fully sponsored transaction block.
     */
    sponsorTransaction(tx) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            return this.request("gas_sponsorTransactionBlock", (0, rpc_js_1.trimTrailingParams)([
                tx.txKind,
                (_a = tx.sender) !== null && _a !== void 0 ? _a : (0, utils_js_1.throwExpression)(new Error("Missing sender")),
                tx.gasBudget,
                tx.gasPrice,
            ]), exports.SponsoredTransaction);
        });
    }
    /**
     * Queries the status of a sponsored transaction block.
     * @param txDigest Sponsored transaction digetst.
     * @returns Sponsored transaction status.
     */
    getSponsoredTransactionStatus(txDigest) {
        return this.request("gas_getSponsoredTransactionBlockStatus", [txDigest], exports.SponsoredTransactionStatus);
    }
    /**
     * Queries the fund associated with the access key.
     * @returns The fund information.
     */
    getFund() {
        return this.request("gas_getFund", [], exports.Fund);
    }
}
exports.GasStationClient = GasStationClient;
/**
 * Builds a gasless transaction.
 * @param txOrBuild Either a `Transaction` object pre-populated with the target transaction data, or
 *    a builder function to populate it.
 * @param options Options
 *    - sender - Optional sender address. Can also be set in the transaction data.
 *    - gasBudget - Optional gas budget. Can also be set in the transaction data.
 *    - gasPrice - Optional gas price. Can also be set in the transaction data.
 *    - sui - `SuiClient`. Required if the transaction uses non fully resolved inputs.
 * @returns A gasless transaction to be sponsored.
 */
function buildGaslessTransaction(txOrBuild, options) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a, _b, _c, _d, _e, _f;
        let tx;
        if (typeof txOrBuild === "function") {
            tx = new transactions_1.Transaction();
            yield txOrBuild(tx);
        }
        else {
            tx = txOrBuild;
        }
        const txData = tx.getData();
        return {
            txKind: (0, utils_1.toB64)(yield tx.build({ client: options === null || options === void 0 ? void 0 : options.sui, onlyTransactionKind: true })),
            sender: (_b = (_a = options === null || options === void 0 ? void 0 : options.sender) !== null && _a !== void 0 ? _a : txData.sender) !== null && _b !== void 0 ? _b : undefined,
            gasBudget: (_d = (_c = options === null || options === void 0 ? void 0 : options.gasBudget) !== null && _c !== void 0 ? _c : txData.gasData.budget) !== null && _d !== void 0 ? _d : undefined,
            gasPrice: (_f = (_e = options === null || options === void 0 ? void 0 : options.gasPrice) !== null && _e !== void 0 ? _e : txData.gasData.price) !== null && _f !== void 0 ? _f : undefined,
        };
    });
}
exports.buildGaslessTransaction = buildGaslessTransaction;
