// @shinami/clients/dist/cjs/sui/zkwallet.js
"use strict";
/**
 * Copyright 2023-2024 Shinami Corp.
 * SPDX-License-Identifier: Apache-2.0
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZkWalletClient = exports.ZkLoginUserId = void 0;
const superstruct_1 = require("superstruct");
const rpc_js_1 = require("../rpc.js");
const utils_js_1 = require("./utils.js");
const endpoints_js_1 = require("./endpoints.js");
const region_js_1 = require("../region.js");
/**
 * Information that identifies a zkLogin user.
 */
exports.ZkLoginUserId = (0, superstruct_1.object)({
    iss: (0, superstruct_1.string)(),
    aud: (0, superstruct_1.string)(),
    keyClaimName: (0, superstruct_1.string)(),
    keyClaimValue: (0, superstruct_1.string)(),
});
/**
 * zkLogin wallet response schema.
 */
const ZkLoginWalletResponse = (0, superstruct_1.object)({
    userId: exports.ZkLoginUserId,
    subWallet: (0, superstruct_1.integer)(),
    salt: (0, superstruct_1.string)(),
    address: (0, superstruct_1.string)(),
});
/**
 * zkLogin wallet RPC client.
 */
class ZkWalletClient extends rpc_js_1.ShinamiRpcClient {
    /**
     * @param accessKey Wallet access key.
     * @param url Optional URL override.
     */
    constructor(accessKey, url = (0, region_js_1.inferRegionalValueFromAccessKey)(accessKey, endpoints_js_1.ZkWalletRpcUrls, (zkWalletRpcUrls) => zkWalletRpcUrls.us1)) {
        super(accessKey, url);
    }
    /**
     * Retrieves a zkLogin wallet or creates a new one if necessary.
     * @param jwt A valid JWT signed by one of the supported OpenID providers.
     * @param keyClaimName The claim name in the JWT that identifies a particular user.
     * @param subWallet The sub-wallet id, which enables the same OpenID user to have more than one
     *    wallet addresses.
     * @returns The zkLogin proof for the given JWT.
     */
    getOrCreateZkLoginWallet(jwt, keyClaimName, subWallet) {
        return __awaiter(this, void 0, void 0, function* () {
            const resp = yield this.request("shinami_zkw_getOrCreateZkLoginWallet", {
                jwt,
                keyClaimName,
                subWallet,
            }, ZkLoginWalletResponse);
            return {
                userId: resp.userId,
                subWallet: resp.subWallet,
                salt: (0, utils_js_1.base64ToBigInt)(resp.salt),
                address: resp.address,
            };
        });
    }
}
exports.ZkWalletClient = ZkWalletClient;
