// @shinami/clients/dist/esm/aptos/gas.js
/**
 * Copyright 2023-2024 Shinami Corp.
 * SPDX-License-Identifier: Apache-2.0
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { AccountAddress, AccountAuthenticator, Deserializer, } from "@aptos-labs/ts-sdk";
import { array, integer, nullable, object, string, unknown, } from "superstruct";
import { ShinamiRpcClient, trimTrailingParams } from "../rpc.js";
import { GasStationRpcUrls, MovementGasStationRpcUrls } from "./endpoints.js";
import { inferRegionalValueFromAccessKey } from "../region.js";
const RpcAccountSignature = object({
    address: string(),
    signature: array(integer()),
});
const SponsorTransactionResult = object({
    feePayer: RpcAccountSignature,
});
const SponsorAndSubmitSignedTransactionResult = object({
    pendingTransaction: unknown(),
});
const Fund = object({
    network: string(),
    name: string(),
    balance: integer(),
    inFlight: integer(),
    depositAddress: nullable(string()),
});
/**
 * Gas station RPC client for Aptos and Movement blockchains.
 * Automatically detects the blockchain from the access key format.
 */
export class GasStationClient extends ShinamiRpcClient {
    /**
     * @param accessKey Gas access key. The format determines which blockchain and network your
     *    transactions are targeting (e.g., "us1_aptos_mainnet_xxx" or "us1_movement_testnet_xxx").
     * @param url Optional URL override.
     */
    constructor(accessKey, url = inferRegionalValueFromAccessKey(accessKey, accessKey.includes("movement")
        ? MovementGasStationRpcUrls
        : GasStationRpcUrls, (gasStationRpcUrls) => gasStationRpcUrls.us1)) {
        super(accessKey, url);
    }
    /**
     * Sponsors a transaction by filling in fee payer address and providing fee payer signature.
     *
     * @param transaction The transaction to request sponsorship for.
     *    Note that `transaction.feePayerAddress` is updated in-place upon successful return.
     * @returns Fee payer signature, to be used in transaction submission.
     */
    sponsorTransaction(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const { feePayer } = yield this.request("gas_sponsorTransaction", trimTrailingParams([
                transaction.rawTransaction.bcsToHex().toString(),
                (_a = transaction.secondarySignerAddresses) === null || _a === void 0 ? void 0 : _a.map((x) => x.toString()),
            ]), SponsorTransactionResult);
            const sig = AccountAuthenticator.deserialize(new Deserializer(new Uint8Array(feePayer.signature)));
            transaction.feePayerAddress = AccountAddress.fromString(feePayer.address);
            return sig;
        });
    }
    sponsorAndSubmitSignedTransaction(transaction, senderSignature, secondarySignatures) {
        return __awaiter(this, void 0, void 0, function* () {
            const secondarySigners = [];
            if (transaction.secondarySignerAddresses) {
                if ((secondarySignatures === null || secondarySignatures === void 0 ? void 0 : secondarySignatures.length) !==
                    transaction.secondarySignerAddresses.length)
                    throw new Error("Unexpected number of secondary signatures");
                secondarySigners.push(...transaction.secondarySignerAddresses.map((address, i) => ({
                    address: address.toString(),
                    signature: secondarySignatures[i].bcsToHex().toString(),
                })));
            }
            const { pendingTransaction } = yield this.request("gas_sponsorAndSubmitSignedTransaction", [
                transaction.rawTransaction.bcsToHex().toString(),
                senderSignature.bcsToHex().toString(),
                secondarySigners,
            ], SponsorAndSubmitSignedTransactionResult);
            return pendingTransaction;
        });
    }
    /**
     * Gets information about the fund associated with the access key.
     */
    getFund() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.request("gas_getFund", [], Fund);
        });
    }
}
